### PR TITLE
CDI, Introduce KUBEVIRT_DEPLOY_CDI flag

### PR DESCRIPTION
**Pull request by oshoval**
In order to make the kubevirt deploy lighter,
and install only whats needed, add a flag that
will determine if to install CDI.
    
This in turn will make less dependencies,
and faster cluster-sync.
    
`export KUBEVIRT_DEPLOY_CDI=false` in order to skip
CDI installation (default is to install it).

For kind providers,  `KUBEVIRT_DEPLOY_CDI` will be set to false automatically,
since we are not using CDI there, and because the loopback devices does conflict with DinD.

Signed-off-by: Or Shoval <oshoval@redhat.com>

```release-note
None
```

### CODE REVIEW

**Comment by oshoval** [comment]
@EdDev @maya-r 
please take a look

**Comment by oshoval** [comment]
Thanks Maya

/test pull-kubevirt-e2e-k8s-1.21-operator

Just because i want to validate the actual number of running tests

**Comment by oshoval** [comment]
/hold

the operator lane need also CDI, but its infra is little different, need to fix first

**Comment by oshoval** [comment]
Please see the suggested fix for sig-opereator here
once its merged, we can simply add the DEPLOY_CDI=true also for the sig-operator, and this way all the lanes interface
will be aligned

https://github.com/kubevirt/kubevirt/pull/6354
https://github.com/kubevirt/project-infra/pull/1547

File: automation/test.sh
Comment (rmohr) [inline automation/test.sh:58]: That should have a `KUBEVIRT_` prefix.

File: hack/cluster-deploy.sh
Comment (rmohr) [inline hack/cluster-deploy.sh:107]: I think that we should add such an option to allow an op-out instead of opt-in to keep things working as they are for everyone right now.

**Review by rmohr — CHANGES_REQUESTED** [review]

**Comment by oshoval** [comment]
Addressed @rmohr comments (testing now)

/hold cancel
since its opt_out now then there is no dependency on the previous mentioned PRs

File: hack/cluster-deploy.sh
Comment (oshoval) [reply hack/cluster-deploy.sh]: Done

File: automation/test.sh
Comment (oshoval) [reply automation/test.sh]: Done

**Comment by oshoval** [comment]
/test pull-kubevirt-e2e-kind-1.19-sriov

errors are not related to PR and known

**Comment by dhiller** [comment]
@oshoval here's my 2cents for integration after kubevirtci KUBEVIRT_DEPLOY_CDI has landed: https://github.com/kubevirt/kubevirt/pull/6375

WDYT?

Also I was not aware of the changes we would have needed for sriov, can you elaborate on those?

**Comment by oshoval** [comment]
> @oshoval here's my 2cents for integration after kubevirtci KUBEVIRT_DEPLOY_CDI has landed: #6375
> 
> WDYT?

I think that its better to keep the logic on cluster-sync, this way there is freedom to update the yamls of CDI easily,
and it would also work on external providers, wdyt?
anyhow, I suggest that we keep both the PRs, lets finish with this one (6339) and then continue with yours ?

> Also I was not aware of the changes we would have needed for sriov, can you elaborate on those?

Sriov did not deploy CDI since a long time ago, because its not needed + because of this comment
https://github.com/kubevirt/kubevirt/pull/6339/files#diff-be559fd8e5c86cd4bfb53d341659b318f4c5963a8da53ca413c19528590edd4bR119
This what you mean ?
the changes that were done are in order to align sriov provider with the ones where `KUBEVIRT_DEPLOY_CDI=false`
but to keep this guard of deleting the `disks-images-provider` (we can remove it as well, keeping in mind kind types must not be deployed with `disks-images-provider`)

**Comment by rmohr** [comment]
> I think that its better to keep the logic on cluster-sync, this way there is freedom to update the yamls of CDI easily,
> and it would also work on external providers, wdyt?

Since the images are prefetched in kubevirtci, we should have the simple knob on `gocli` to deploy it or not.  That does never rule out that someone just chooses to not deploy the default CDI and just deploys a random manifest to get different versions.

**Comment by oshoval** [comment]
> Since the images are prefetched in kubevirtci, we should have the simple knob on `gocli` to deploy it or not. That does never rule out that someone just chooses to not deploy the default CDI and just deploys a random manifest to get different versions. That is planned for a long time now.

If someone will deploy a custom random manifest, he will need to undeploy it manually every time before a secondary cluster-sync
else he cant undeploy kubevirt

Anyhow, if possible that we will proceed with this PR, and then continue with the PRs of Daniel i will be glad,
this one was tested and its a standalone change that improves cluster-sync timing.

**Comment by rmohr** [comment]
> If someone will deploy a custom random manifest, he will need to undeploy it manually every time before a secondary cluster-sync
> else he cant undeploy kubevirt (so it makes it more natural on cluster-sync imho).

How often do you need to switch CDI versions during development? I think this is a mostly theoretical discussion.

> 
> Anyhow, if possible that we will proceed with this PR, and then continue with the PRs of Daniel i will be glad,
> this one was tested and its a standalone change that improves cluster-sync timing, without any regression.

I don't mind if that gets in in the meantime.

**Comment by oshoval** [comment]
> > If someone will deploy a custom random manifest, he will need to undeploy it manually every time before a secondary cluster-sync
> > else he cant undeploy kubevirt (so it makes it more natural on cluster-sync imho).
> 
> How often do you need to switch CDI versions during development? I think this is a mostly theoretical discussion.

Maybe the CDI people do need it when they are debugging,
But if its not interesting case, then np.
Another point here is that undeploy is needed even if there isn't a swich, commented about it on the ref PR.

> 
> > Anyhow, if possible that we will proceed with this PR, and then continue with the PRs of Daniel i will be glad,
> > this one was tested and its a standalone change that improves cluster-sync timing, without any regression.
> 
> I don't mind if that gets in in the meantime.

Awesome, will be glad for reviews please then.

**Comment by rmohr** [comment]
> Maybe the CDI people do need it when they are debugging,
> But if its not interesting case, then np.

You can still deploy anything you want for debugging. Deploying different CDI versions directly via `kubectl apply` from github is potentially even easier for trying out different versions than messing around with the checked in manifests.

> Another point here is that undeploy is needed even if there isn't a swich, commented about it on the ref PR.

No.

File: hack/cluster-deploy.sh
Comment (dhiller) [inline hack/cluster-deploy.sh:120]: Wouldn't this be unnecessary in case of `KUBEVIRT_DEPLOY_CDI=false` as we are already removing the manifest here?
https://github.com/kubevirt/kubevirt/pull/6339/files#diff-be559fd8e5c86cd4bfb53d341659b318f4c5963a8da53ca413c19528590edd4bR38

File: hack/cluster-deploy.sh
Comment (dhiller) [inline hack/cluster-deploy.sh:105]:
```suggestion
# TODO: Remove it when cdi is supported on ARM
if [[ $KUBEVIRT_DEPLOY_CDI == true ]] && [[ ${ARCHITECTURE} != *aarch64 ]]; then
```

File: automation/test.sh
Comment (dhiller) [inline automation/test.sh:49]: Did you check that no additional tests were skipped when adding this?

File: automation/test.sh
Comment (dhiller) [inline automation/test.sh:54]: Did you check that no additional tests were skipped when adding this?

File: automation/test.sh
Comment (oshoval) [reply automation/test.sh:49]: yes, all is good, the only skipped were on the sig-operator, which does have KUBEVIRT_DEPLOY_CDI=true
so no problem there

File: automation/test.sh
Comment (oshoval) [reply automation/test.sh:54]: yes, all is good, the only skipped were on the sig-operator, which does have KUBEVIRT_DEPLOY_CDI=true
so no problem there

File: hack/cluster-deploy.sh
Comment (oshoval) [reply hack/cluster-deploy.sh:120]: we have several kind types
```
kind-1.19-sriov
kind-1.19-vgpu
kind-k8s-1.17
```
only for the sriov we are setting KUBEVIRT_DEPLOY_CDI=false
so its a global guard for kind type, as the condition here is kind and not sriov or such
( we can set it for all kind types here if needed
https://github.com/kubevirt/kubevirt/pull/6339/files#diff-ce25e369a9df9edc525425cc144fc8c40599fff3cc1e5058b64336d71af67162R74 )

**Comment by oshoval** [comment]
first push - address comments
2nd push - rebase

**Comment by oshoval** [comment]
`06:45:34: binding device to vfio driver failed`
not related

/test pull-kubevirt-e2e-k8s-1.20-operator-nonroot

**Comment by oshoval** [comment]
/retest

not related

**Comment by rmohr** [comment]
/approve

@dhiller maybe you can have a final look since you looked at this from another perspective already and give a lgtm if it looks ok.

**Review by dhiller — APPROVED** [review]

**Comment by oshoval** [comment]
Thanks
should we add to hack/cluster-deploy.sh that in case the variable doesn't exist its value will be true
in the very beginning ? this way it will be opt-out for the people who use kubevirt locally and used to have CDI and needs it
(he can also `export KUBEVIRT_DEPLOY_CDI=true` but just on the first time will need to know about it)

**Pull request merged**
